connectors/kafka/src/main/java/forklift/controller/KafkaController.java

```diff
@@ -212,39 +238,59 @@ private void controlLoop() {
         }
     }
 
-    private void commitPendingAndWaitForTopics() throws InterruptedException {
-        //check if the last remaining topic was removed
-        if (kafkaConsumer.assignment().size() > 0) {
-            Map<TopicPartition, OffsetAndMetadata>
-                            offsets =
-                            acknowledgmentHandler.removePartitions(kafkaConsumer.assignment());
+    private boolean processTopicChanges() throws InterruptedException {
+        boolean topicsChanged = processTopicAdd();
+        topicsChanged = processTopicRemoved() || topicsChanged;
+        if (topicsChanged) {
+            kafkaConsumer.subscribe(topics, new RebalanceListener());
+        }
+        return topicsChanged;
+    }
+
+    private boolean processTopicAdd() {
+        if (topicToAdd != null) {
+            topics.add(topicToAdd);
+            messageStream.addTopic(topicToAdd);
+            topicToAdd = null;
+            return true;
+        }
+        return false;
+    }
+
+    private boolean processTopicRemoved() throws InterruptedException {
+        if (topicToRemove != null) {
+            topics.remove(topicToRemove);
+            messageStream.removeTopic(topicToRemove);
+            Set<TopicPartition> removed = kafkaConsumer.assignment().stream().filter(partition -> !topics.contains(partition.topic())).collect(Collectors.toSet());
```

Comment:
Is it not equivalent to check `partition.topic() == topicToRemove` instead of `!topics.contains(partition.topic())` here?